### DISCUSSION TITLE
Decide if the airstream sender should be the Airstream admin

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/v2-periphery#155] [Ideas]
Airstream contracts will inherit from [`Adminable`](https://github.com/sablier-labs/v2-core/blob/fcbc567aea6362c1f102eb19732badeb4899dec5/src/abstracts/Adminable.sol), and there comes the question of who the stream sender should be when claiming airstreams?

It can be either the Airstream contract itself or the administrator's address.

The massive benefit of setting the admin as the stream sender would be that we wouldn't have to implement `cancel` and `cancelMultiple` in the Airstream contract. The typical cancellation flow would apply.

The only con I can think of is that this might interfere the accounting logic since the nominal sender would be different from the real sender (the Airstreams contract). However, this should be a negligible issue - many airstreams will likely be non-cancelable, and in any case the sender and the funder are different entities in V2 Core:

https://github.com/sablier-labs/v2-core/blob/fcbc567aea6362c1f102eb19732badeb4899dec5/src/interfaces/ISablierV2LockupLinear.sol#L28-L38

**Comment by razgraf** [comment]
I like the idea of having the Airstream contract be the sender. 

- It would make the system self contained
- It simplifies the logic of funds "traveling" through multiple accounts, which allows airdrop funds to only be tracked through the Airstream contract (as opposed to balances in the contract and in the sender's wallet). Meaning token entries in the sender's wallet would only be done as a result of a manual clawback.

However, I do see how this would lower the burden of having to implement extra methods.

Informational note: regarding the accounting flow, the proxy is also going to create special "hoops". Not that this is any different from how the other contracts work at the moment. I assume, at least at the app level, we'll ask the sender to first create a proxy. Through it they'll deploy an Airstream campaign with the factory. Then, the cancelation transfers become:

| Flows |       | 
| :----- | --- |
| Sender = Airstream | `Lockup -> Airstream` |
| Sender = User | `Lockup -> Proxy -> Proxy Owner` |

Begs a separate question for our design: do we rely on sender proxies for this at all?

**Reply by razgraf** [reply]
Side thought: I know we should define a clear "recommended" way to do this - so the conversation is still valuable. But can't we make this `sender` a parameter in the factory method similar to what we already to in the [lockup](https://github.com/sablier-labs/v2-core/blob/fcbc567aea6362c1f102eb19732badeb4899dec5/src/SablierV2LockupLinear.sol#L274) contracts?

**Reply by PaulRBerg** [reply]
Unfortunately, having the AC (Airstream contract) be the stream sender would neither make the system self-contained nor would it simplify the transfer logic, at least given the current intended design. This is because the cancel functions would auto-forward the refunded assets to the admin. We can't rely on `clawback` because that function is gated by the `expiration`. Canceled funds would be locked until the airdrop expires, which wouldn't make sense from a capital efficiency point of view.

Then, regarding the proxy. That is not included in how I currently envision the airstreams' architecture. I've mulled this topic over since our call on Saturday, and the proxy design doesn't work here because we need (at least) one storage property, i.e., the `claimed` mapping that specified which users have claimed the airstreams.

> But can't we make this sender a parameter in the factory method similar to what we already to in the [lockup](https://github.com/sablier-labs/v2-core/blob/fcbc567aea6362c1f102eb19732badeb4899dec5/src/SablierV2LockupLinear.sol#L274) contracts?

We could but it seems overkill to me since:

1. We would introduce several if statements, e.g. to check that the `sender` is not the airstreams contract itself, the zero address, etc.
2. The end goal is the same - to auto-forward the assets to the admin.

**Reply by PaulRBerg** [reply]
Alternatively, we could implement yet another function like `withdrawRefunds` and yet another state variable `refundedAmount`, but doing so seems overkill considering that it would introduce an additional accounting flow. Users would need to track (i) unclaimed airstreams and (ii) canceled balances. Additionally, the security of the contract would be impaired by way of there being 3, 4 extra functions, one of which would be a withdraw function.

I'm leaning toward making the sender the admin and call it a day.

**Reply by razgraf** [reply]
Proxy: wasn't proposing a proxy design for the factory. I was talking about the sender itself interacting with the factory through a proxy, like they do now with the lockup contracts. Would make it "consistent" with the lockup system and would future proof us in case we ever need "complex" actions, not natively coded inside the factory itself. There's a case to be made for YAGNI here, but still, worth bringing up.

**Reply by razgraf** [reply]
Good point with the cancel vs clawback balances. Still, I'm more attracted to the idea of bringing the contract in a "self-contained" state, even with separate balances as you said (a.k.a. custom logic).

**Reply by PaulRBerg** [reply]
> Proxy: wasn't proposing a proxy design for the factory. I was talking about the sender itself interacting with the factory through a proxy

The latter sentence is precisely what it means to have a proxy design for the factory. If you want the sender to interact with the factory through a proxy, the factory becomes a target contract.

> would future proof us in case we ever need "complex" actions, not natively coded inside the factory itself.

We can just write a new factory and deploy it. Airstream creators are not "connected" to the factory in any way.

And the `Airstream` contract cannot be a target because airdrop recipients interact with it.

> a case to be made for YAGNI here

Yes, big case :D

> I'm more attracted to the idea of bringing the contract in a "self-contained" state

Can you put your finger on the precise reasons why you prefer the self-contained design? It would significantly increase the amount of logic in the `Airstream` contract:

- `cancel`
- `cancelMultiple`
- `getRefundedAmount`
- `withdrawRefundedAmount`

Additionally:

- it would make the implementation of `clawback` much more complicated
- it would require the admin to monitor two balances rather than just one
- it would increase the gas costs
- it would increase the audit cost
- it would be a deadweight balance because the refunded assets couldn't be used for a new airstream campaign. The new campaign would have to be deployed via the factory (assuming we go with option 1 discussed here https://github.com/sablier-labs/v2-periphery/discussions/156)

**Reply by razgraf** [reply]
Thank you for the reply. I think you're right. Technically, having the Sender and the Airstream entities separate would decrease complexity, therefore we could have a faster go-to-market sprint.

The self-contained idea - was thinking that for the sender, an "Airstream" contract could act as a `facade`, such that they wouldn't have interact much with the endpoint contracts (lockup linear, dynamic).

The accounting (mental) complexity remains: for a DAO, we'd probably suggest they use a dedicated wallet when creating and funding the airstream contract, not their main treasury. If for some reason people start canceling their streams (or they as senders have to, in multiple batches due to gas restrictions with createMultiple), they'd have an unwanted wave of entries in their treasury wallet. 

Side note: Would be worth writing a guide for "best practices" with airstreams once we have it ready.

--

TL;DR; Let's try with `Sender = Owner, not Airstream` approach and see what users/auditors suggest as well. @andreivladbrg @gavriliumircea any thoughts on the above?

**Reply by gavriliumircea** [reply]
I do understand both your points., but right now it seems to me that we would benefit more from making the sender the admin.

**Reply by PaulRBerg** [reply]
Thanks!

> that they wouldn't have interact much with the endpoint contracts (lockup linear, dynamic).

They will be familiar with `LockupLinear`/ `LockupDynamic`.

1. All of their users will need to interact with those contracts
2. It is likely that the campaign creators themselves will have some airstreams to claim and later withdraw from Sablier

I can also envision some airstream creators that build customized UIs for withdrawing from Sablier.

> we'd probably suggest they use a dedicated wallet when creating and funding the airstream contract, not their main treasury.

I wouldn't go that far now. The next steps are (i) building the system, (ii) auditing it, and (iii) presenting it to users. Only after the 3rd step can we validate accounting best practices.

> they'd have an unwanted wave of entries in their treasury wallet.

This is always possible on the blockchain - Sablier does not alter the status quo in any significant way.

> Would be worth writing a guide for "best practices" with airstreams once we have it ready

Later on, yes. This would go on [docs.sablier.com](https://docs.sablier.com).

> Let's try with `Sender = Owner, not Airstream` approach

Sounds good

**Reply by andreivladbrg** [reply]
I see that you've reached an agreement on `sender = admin`. I also thought this approach to be the better one, I don't have anything to add at the moment.

**Reply by PaulRBerg** [reply]
cheers @andreivladbrg.
